File: src/apps/MultitouchExtension/src/FingerStatusEntry.m
#import "FingerStatusEntry.h"

@implementation FingerStatusEntry

- (instancetype)initWithDevice:(MTDeviceRef)device
                    identifier:(int)identifier {
  self = [super init];

  if (self) {
    _device = device;
    _identifier = identifier;
    _frame = 0;
    _point = NSMakePoint(0, 0);
    _touchedPhysically = NO;
    _touchedFixed = NO;
    _ignored = YES;
    _delayTimer = nil;
  }

  return self;
}

- (instancetype)copyWithZone:(NSZone*)zone {
  printf("make copy\n");

  FingerStatusEntry* e = [[FingerStatusEntry alloc] initWithDevice:self.device identifier:self.identifier];

  e.frame = self.frame;
  e.point = self.point;
  e.touchedPhysically = self.touchedPhysically;
  e.touchedFixed = self.touchedFixed;
  e.ignored = self.ignored;
  // e.delayTimer is nil

  return e;
}

@end

File: src/apps/MultitouchExtension/src/FingerStatusEntry.swift
enum FingerStatusEntryTimerMode {
  case none
  case touched
  case untouched
}

class FingerStatusEntry: Identifiable {
  public var id = UUID()

  //
  // Unique keys
  //

  var device: MTDevice
  var identifier = 0

  //
  // Variables
  //

  var frame = 0

  var point = NSMakePoint(0, 0)

  // True if the finger is touched physically.
  var touchedPhysically = false

  // True if the finger is touched continuously for the specified time. (finger touch detection delay)
  var touchedFixed = false

  // True while the finger has never entered the valid area.
  var ignored = true

  /*
@property NSTimer* delayTimer;
@property enum FingerStatusEntryTimerMode timerMode;
*/

  //
  // Methods
  //

  init(device: MTDevice, identifier: Int) {
    self.device = device
    self.identifier = identifier
    /*
    _delayTimer = nil
    _timerMode = FingerStatusEntryTimerModeNone
    */
  }
}
